libraries/AP_HAL_Linux/Poller.cpp

```diff
@@ -0,0 +1,403 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include <algorithm>
+#include <arpa/inet.h>
+#include <cstddef>
+#include <errno.h>
+#include <fcntl.h>
+#include <netinet/in.h>
+#include <netinet/tcp.h>
+#include <pthread.h>
+#include <stdint.h>
+#include <sys/epoll.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "Poller.h"
+#include "Scheduler.h"
+
+extern const AP_HAL::HAL& hal;
+
+namespace Linux {
+
+uint32_t Poller::to_epoll_events(Poller::Event ev) {
+    // EPOLLWAKEUP prevents the system from hibernating or suspending when
+    // inside epoll_wait() for this particular event.  It is silently
+    // ignored if the process does not have the CAP_BLOCK_SUSPEND
+    // capability.
+    uint32_t op = EPOLLWAKEUP;
+
+    if (ev & Poller::Event::Read) {
+        op |= EPOLLIN;
+    }
+    if (ev & Poller::Event::Write) {
+        op |= EPOLLOUT;
+    }
+    if (ev & Poller::Event::Error) {
+        op |= EPOLLERR;
+    }
+
+    return op;
+}
+
+bool Poller::register_pollable(Pollable *p, const Poller::Event ev) {
+    if (_epfd < 0) {
+        return false;
+    }
+
+    struct epoll_event epev = {0};
+    epev.events = Poller::to_epoll_events(ev);
+    epev.data.ptr = static_cast<void *>(p);
+
+    return epoll_ctl(_epfd, EPOLL_CTL_ADD, p->get_fd(), &epev) == 0;
+}
+
+void Poller::unregister_pollable(const Pollable *p) {
+    if (_epfd >= 0) {
+        epoll_ctl(_epfd, EPOLL_CTL_DEL, p->get_fd(), NULL);
+    }
+}
+
+int Poller::poll(int timeout_ms) const {
+    const int max_events = 16;
+    epoll_event events[max_events];
+    auto before_wait = AP_HAL::millis();
+    int r = epoll_wait(_epfd, events, max_events, timeout_ms);
+
+    if (r > 0) {
+        const auto delta = AP_HAL::millis() - before_wait;
+        const auto max_time_ms = std::max((timeout_ms - delta) / r, 1U);
+
+        for (int i = 0; i < r; i++) {
+            Pollable *p = static_cast<Pollable *>(events[i].data.ptr);
+
+            if (events[i].events & EPOLLIN) {
+                p->on_can_read(max_time_ms);
+            }
+            if (events[i].events & EPOLLOUT) {
+                p->on_can_write(max_time_ms);
+            }
+            if (events[i].events & EPOLLERR) {
+                p->on_error(max_time_ms);
+            }
+            if (events[i].events & EPOLLHUP) {
+                p->on_hang_up(max_time_ms);
+            }
+        }
+    } else if (r < 0) {
+        if (errno == EINTR) {
+            // A possible nudge from another thread; return 0 to signal
+            // tmout.
+            r = 0;
+        }
+    }
+
+    return r;
+}
+
+Pollable::~Pollable() {
+    auto scheduler = Linux::Scheduler::from(hal.scheduler);
+    scheduler->unregister_pollable(this);
```

Comment:
What if there are methods such as `Pollable::is_real_time()`, or `Pollable::get_scheduler_hint()` returning an `enum { HINT_RT, HINT_NONRT, ... }`, that's used by the scheduler to know which thread to add that pollable?

Another alternative would be passing the scheduler hint as a parameter to the `register_pollable()` method. There's no need to pass the same hint to the `unregister_pollable()` method since, whenever a file descriptor is closed, it's removed from whatever epoll set it was present.